### DISCUSSION TITLE
Ask FP anything!

### DISCUSSION THREAD

**Discussion by Zaydme** [FightPandemics/FightPandemics#2064] [Q&A]
# :question:  Ask FP anything!

**Comment by ddveloper** [comment]
Hi @Zaydme , I have a question about `Promise` and `resolve`, in PR #1995 the new `delete` endpoint shares 95% of code with the `post` endpoint, so I tried to extract the common logic out of those 2 and add `Promise` and `resolve` to fulfill the promise but got Fastify error like below, any ideas how to fix this?
```
backend-service    | [2020-12-09 12:16:51.883 +0000] INFO  (405 on bae19b3161db): incoming request
backend-service    |     reqId: 1
backend-service    |     req: {
backend-service    |       "method": "POST",
backend-service    |       "url": "/api/users/current/avatar",
backend-service    |       "hostname": "backend-service:8000",
backend-service    |       "remoteAddress": "172.18.0.4",
backend-service    |       "remotePort": 49546
backend-service    |     }
backend-service    | [2020-12-09 12:16:52.365 +0000] ERROR (405 on bae19b3161db): Promise may not be fulfilled with 'undefined' when statusCode is not 204
backend-service    |     reqId: 1
backend-service    |     err: {
backend-service    |       "type": "FastifyError",
backend-service    |       "message": "FST_ERR_PROMISE_NOT_FULLFILLED: Promise may not be fulfilled with 'undefined' when statusCode is not 204",
backend-service    |       "stack":
backend-service    |           FastifyError [FST_ERR_PROMISE_NOT_FULLFILLED]: FST_ERR_PROMISE_NOT_FULLFILLED: Promise may not be fulfilled with 'undefined' when statusCode is not 204
backend-service    |               at /app/node_modules/fastify/lib/wrapThenable.js:34:30
backend-service    |               at processTicksAndRejections (internal/process/task_queues.js:97:5)
backend-service    |       "name": "FastifyError [FST_ERR_PROMISE_NOT_FULLFILLED]",
backend-service    |       "code": "FST_ERR_PROMISE_NOT_FULLFILLED",
backend-service    |       "statusCode": 500
backend-service    |     }
localstack         | 2020-12-09 12:16:52,335:API: 172.18.0.5 - - [09/Dec/2020 12:16:52] "PUT /fp-dev-cdn/content/avatars/users/5fd09ff7aa061b110089a41c-1607516212283 HTTP/1.1" 200 -
```
my changed code is attached, thank you 🍪 
[users.js.txt](https://github.com/FightPandemics/FightPandemics/files/5665779/users.js.txt)

**Reply by Zaydme** [reply]
@ddveloper 
fasify says ` Promise may not be fulfilled with 'undefined'` when you are not returning any response, or when it is undefined (which is only allowed if status Code is 204, which is "No Content")

When the promise is resolved here
![2](https://user-images.githubusercontent.com/46847212/101630484-37d67500-3a23-11eb-969b-e1cafabbd55a.png)
you should return that value here, now you just wait for it and do nothing with it, and the request ends without response
![code](https://user-images.githubusercontent.com/46847212/101630280-f940ba80-3a22-11eb-9ddd-16f0eeac425a.png)

you will need to do something like this
![2](https://user-images.githubusercontent.com/46847212/101630668-808e2e00-3a23-11eb-8994-c96e594389f0.png)

**Reply by ddveloper** [reply]
Thanks @Zaydme 👍 you save me another day 🌞

**Reply by ddveloper** [reply]
@Zaydme another question is when there are `throw app.httpErrors.internalServerError();` in the promise function, do I `reject` for those `throw`s? I noticed when no reject and error was thrown, it reports a scary 😱  warning:
```
[2020-12-09 13:04:28.018 +0000] ERROR (89 on bae19b3161db): Failed updating organisation avatar.
backend-service    |     ReferenceError: user is not defined
backend-service    |         at /app/lib/endpoints/organisations.js:405:9
backend-service    |         at processTicksAndRejections (internal/process/task_queues.js:97:5)
backend-service    | (node:89) UnhandledPromiseRejectionWarning: InternalServerError: Internal Server Error
backend-service    |     at Object.internalServerError (/app/node_modules/fastify-sensible/lib/httpErrors.js:137:19)
backend-service    |     at /app/lib/endpoints/organisations.js:451:30
backend-service    |     at processTicksAndRejections (internal/process/task_queues.js:97:5)
backend-service    | (node:89) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
backend-service    | (node:89) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
backend-service    | [nodemon] restarting due to changes...
```

**Reply by Zaydme** [reply]
@ddveloper can you share `/lib/endpoints/organisations.js`

**Reply by ddveloper** [reply]
[organisations.js.txt](https://github.com/FightPandemics/FightPandemics/files/5666248/organisations.js.txt)
note the `user` in line 405 is intentional error to trigger the 500

**Reply by Zaydme** [reply]
i don't know if this will work, but it is worth trying,
instead of doing this:
![2](https://user-images.githubusercontent.com/46847212/101638693-a7059680-3a2e-11eb-8aa2-6e89feaf2d26.png)
reject with something like this 
```js
    reject(app.httpErrors.internalServerError())
```
(I don't know what `app.httpErrors.internalServerError()` does/returns exactly so I doubt if this will work, but worth trying)

and then catch it and throw it.
![2](https://user-images.githubusercontent.com/46847212/101638962-06fc3d00-3a2f-11eb-9537-4d69d2da6171.png)

**Reply by ddveloper** [reply]
Yes, that works 👍 
```
backend-service    | [2020-12-09 14:11:25.682 +0000] ERROR (46 on bae19b3161db): Failed updating organisation avatar.
backend-service    |     ReferenceError: user is not defined
backend-service    |         at /app/lib/endpoints/organisations.js:405:9
backend-service    |         at processTicksAndRejections (internal/process/task_queues.js:97:5)
backend-service    | [2020-12-09 14:11:25.684 +0000] INFO  (46 on bae19b3161db): request completed
backend-service    |     reqId: 5
backend-service    |     res: {
backend-service    |       "statusCode": 500
backend-service    |     }
backend-service    |     responseTime: 37.11199999973178
```
P.S. the internalError is handled like this:
![image](https://user-images.githubusercontent.com/44758313/101640851-e360d480-39e5-11eb-9a39-fa7eb8823127.png)
